### DISCUSSION TITLE
Sidekick64 or RAD?

### DISCUSSION THREAD

**Discussion by Ratteler** [frntc/RAD#11] [General]
Obviously I don't think I'll be able to plug in both or these. 17xx and GEOS compatibility is on the top of list, but it seems like the Sidekick does everything BUT 17xx compatibility. So I'm really torn as to which one to go for.

Any chance of feature migration? Can the Sidekick work with Pi Zero 2W and do 17XX emulation in a future update?

Will the RAD get some more of the SideKick feature in the future?

What is the technical differences between the 2 devices?

**Comment by frntc** [comment]
Get both of course :-)

(technical differences: see answer in SK64 discussions -> glue logic to interface RPi and C64)

I'm thinking of / looking into adding more features to RAD, but it will not be able to do the same as Sidekick64 does.

I was also considering enabling parallel use of the two, although that seems quite a bit exaggerated :-)

**Comment by Ratteler** [comment]
If they would work together I would get both! But if my memory of cartridge port from my days as CMD beta tester a correct, I would think the overlap would bork something.

Would D64/D71/D81 comparability be possible with the RAD?

I suppose a Sidekick RAD is not in the cards. LOL.

**Reply by frntc** [reply]
Regarding disk images (copied from the other thread):

About the D64-support: if it's about starting a PRG from a D64-image -- this would be diligence work only. If you're talking about disk emulation (even via kernal vectors) that's a different story as the RAD is not meant for this. It would require the same approach as the Turbo Chameleon (emulating the CPU plus everything that you want)

**Reply by Ratteler** [reply]
Is there a way to save GEOS files to the SD Card? That's going to be my #1 use for all that REU space.

**Reply by frntc** [reply]
I really don't know much about GEOS, but I think I succesfully tested a RAM-disk in the REU-emulation which I saved to SD.

**Comment by webmst** [comment]
I think it is anyway advisable to conceive (and publish) an updated standard framework to connect a RP to a C64/128 expansion port, in the same original intent of your RASPIC64 project, that you declared "superseded" just because - in order to implement "more functions" for the SIDEKICK project you thought of a different scheme to allocate inputs and outputs on the RP.

IMHO, after having analyzed your work, I propose the RAD scheme as the most functional to be used as a general-purpose framework.

 To control ALL the signals present on the expansion port it just lacks the ROML input (that you can read connecting it to pin5 of the 74257 currently connected to a testpoint) and the EXROM output that can be controlled by the pin on the RP currently used for the 74245's DIR after adapting the voltage level with one of the many available ports in the 743861. The direction for the '245 can now be obtained by a simple EXOR (IC:7486) of DMA and R/W signals (in few words R/W sets the direction for the data bus, that is reversed in case of a DMA). Let me say that all the other possibile functions for the RP (e.g. display management, more buttons) are less important and can be replaced by an OSD on the C64.

**Reply by frntc** [reply]
RASPIC64 is a subset of Sidekick64, that's why I declared it superseded. I really kept it just for the sake of development history.

You're right that the 257 has one spare input which could read ROML. I prefer not letting RW be controlled by some logic on the board which has its benefits, e.g. when OE sets the 245 active, GPIOs are set to output and RW is wrong then there's a clash which will probably kill your GPIOs.

But even if I'd save on GPIO (and getting back to the core of your question): the RAD is not the most functional -- it can do DMA and IO1/2, but it can't emulate ROML/ROMH cartridges without limitations as A8..15 are connected to a LVC573.

**Reply by webmst** [reply]
Speaking about the '245 DIR signal, I understand your concern of ensuring protection of the RP GPIO from a bus conflict still I am completely confident you would avoid it by disabling its CE (in that case BOTH the A and the B bus of the '245 are set to isolation when high) if the data bus is not meant to be accessed, no matter which direction is (automatically) set; anyway it's the only contrivance I devised to save an extra output pin for EXROM.

Thus, regarding the address bus detection (as output is already set&latched by the '573) and completely agreeing with you on maintaining feature compatibility for the SIDEKICK, I implied (sorry if I didn't make it explicit in my post) you would have completed the current RAD design with an "input" circuitry for the address bus (e.g. 2x74244), driven by an inverted OE for the '573. This could eventually spare a '257 for the detection of the LSB and a 7430 for a (non-hardwired but software-defined) detection of the MSB on the address bus, providing ROML/ROMH emulation.

I used a similar bidirectional approach (but without multiplexing address&data busses on a single shared 8-bit port) in another project of mine, so please let me know if I am overlooking something about your design.

**Reply by frntc** [reply]
You mean using a LVC573 for output and "parallel" to is a LVC244/245 as input? I was thinking about something like this, but there are times when the 573 is disabled and the 244/245 must not be enabled. Keep in mind that the same 8 GPIOs output A0..7/A8..15/D0..7 and input D0..7 -- and I think then it won't work if I understood you correctly. A way to go for the design might be bidirectional multiplexers, but these ICs are a bit less common and I haven't experimented with them. 

One more historical reason for controlling RW explicitly: on the Sidekick64 the D0..7-LVC245 is controlled by the bus RW which works great on a C64. But the C128 does not set RW correctly during Ultimax VIC-accesses...

**Reply by webmst** [reply]
Full-blown schematics are worth one thousand words, so I'm attaching my proposal here.
It encompasses all the modifications we've addressed before, getting straight to the final result so you can see the full picture; my design still maintains compatibility with your original design (please check).

BTW, I have never seen a pre-made circuit for a "bidirectional tristatable buffer w/ output latch", so I've always built something like this in all my projects.

[rad-0.2a.pdf](https://github.com/frntc/RAD/files/11242004/rad-0.2a.pdf)

P.S.
What do you mean with "But the C128 does not set RW correctly during Ultimax VIC-accesses"?
In the C64/128 the R/W signal is always output by CPU(s) as for any other I/O device; it is an input for both the C64 and C128 PLAs (i.e. the signal is unmodified, no matter GAME/EXROM status) and for the VIC-IIe R/ W is just used as an input for register access, while during a DMA the CPU sets R/W to HiZ, leaving it pull-upped (read). So for a VIDEO DMA (BA low) VIC always reads, unless an external DMA (allowed only when BA and PHI2 are high) pulls the signal otherwise.

**Reply by frntc** [reply]
I'll have a closer look later (no time atm). Btw. I meant the same with 573-parallel-to-245 as you have in your schematics. 

If I get your schematics right on first glance there might be a few things to consider:
- you could not handle Ultimax VIC accesses (because BA masks OEs), right? 
- once the VIC pulls BA you have 3 more cycles to write to the bus (in fact the original REU uses one of these cycles in a special case when BA is a bit late)
- not having NMI as output would not allow the emulation of freezer cartridges
- having Phi2 only on a multiplexer does not allow you to use FIQs for bus communication (SK64 almost exclusively uses this, RAD for some code paths)
- some nasty details I'd have to reconstruct from the SK64 myself: the shuffling of signals to multiplexers depends on when signals are available on the bus and how much/little time you have to mem access/cache preload

In essence, maybe the 573-parallel-to-245 setup works, but I never tried as getting things right was so much experimentation that I wanted to reuse the SK64 experience. But hey, give it a try! (The core of the bus communication is a pretty small part of the source code.)

C128 and RW: I would have to dig out the reference (some forum discussion), but from SK64 I can tell -- if you use RW to set the 245-direction it will not work for every cycle. And when the VIC pulls BA it could still be up to 3 write cycles from the CPU! (see AEC signal and how to go into DMA mode)

**Reply by webmst** [reply]
In my schematics - as a starting point - I have tried to keep your original RAD setup and develop it as per my initial choice.

In this first checking instance we are assessing if we can spare enough hw resources to fully control the expansion bus with a unified design. Some consequent issues, expecially on the RP side, were expected but I cannot anticipate too much without teaming up with you.

In this regard, let me say you have all my engineering respect: I know how much it is difficult to make software to emulate hardware in realtime, even designing a 1MHz application using a bare-metal development environment on a multi-core SBC running at 1.3GHz.  

Back to the issues:

- NMI was already read via mux in RAD, so it was input-only; I simply wrote a reminder to address it later
- I had to move PHI2 (input) to the mux in order to drive EXROM (output); I was expecting some timing issues software-side

_We can discuss if there's another signal worth sacrificing for NMI: given the importance of the freezer function I would vote for the RES signal, as there's always a manual RESET button available (simply swapping NMI and RES we could read the reset signal to notify your software), but you may think otherwise. Let me know.
Speaking about PHI2 and the FIQ interrupt, please tell me if you can modify the software to work with the new GPIO arrangement, as I don't see any other options._

Regarding BA and OEs, I have simply kept the same mechanism you had implemented with the previous 8-input NAND GATE (74HC30).

Speaking about the DMA specification (that is the one I reported at the end of my previous post), I was describing an EXTERNAL DMA SOURCE, that should strictly wait for PHI2 and BA high before accessing the bus and that should release DMA immediately when BA goes low (monitoring BA and keeping the data transfer in sync with PHI2 in software is fundamental here). Let me add that the VIC-IIe encompasses two (unused) pins: DMAREQ and DMAACK that implement an early priority-based DMA controller; it could acknoledge an external DMA request via the DMAREQ pin when BA and PHI2 are high through the DMAACK pin, still assigning priority to the VIC itself (I suppose it was for a future upgrade or to spare some glue logic in a later C128 board release).

**Reply by webmst** [reply]
I thought of another alternative to free some "direct" pins on the RP: to connect the mux outputs to the common parallel I/O port, enabling its OE when needed with one of the spared pins (e.g. GPIO3). In this way, 3 pins (GPIO0-2) would be available; one (e.g. GPIO2) would solve the previous NMI issue (leaving a mux input available for something else, e.g. another button) and two other RP pins (GPIO0-1) would still be available for other functions (e.g. a I2C display for the sidekick). Just let me know if these options are doable in software.
